# Pratishtha Singh- 8927578830

### 1. Multi-class and Multi-Label Classification Using Support Vector Machines- (a)

In [236]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings(action='ignore')
from sklearn import svm
from sklearn.metrics import hamming_loss
from sklearn.metrics import accuracy_score
df=pd.read_csv("Frogs_MFCCs.csv")
df['Family']=df['Family'].map({'Bufonidae':'0','Dendrobatidae':'1','Hylidae':'2','Leptodactylidae':'3'})
df['Genus']=df['Genus'].map({'Adenomera':'0','Ameerega':'1','Dendropsophus':'2','Hypsiboas':'3','Leptodactylus':'4',
                             'Osteocephalus':'5','Rhinella':'6','Scinax':'7'})
df['Species']=df['Species'].map({'AdenomeraAndre':'0','AdenomeraHylaedactylus':'1','Ameeregatrivittata':'2','HylaMinuta':'3',
                                'HypsiboasCinerascens':'4','HypsiboasCordobae':'5','LeptodactylusFuscus':'6',
                                 'OsteocephalusOophagus':'7','Rhinellagranulosa':'8','ScinaxRuber':'9'})
data_x=df.iloc[:,:22]
data_y=df.iloc[:,22:25]
x_train, x_test, y_train, y_test=train_test_split(data_x, data_y, test_size=0.3)

### 1. Multi-class and Multi-Label Classification Using Support Vector Machines- (b) (i)

Hamming Loss is the fraction of labels incorrectly predicted.

Exact Match is used to find the subset accuracy in multilabel classification. The set of labels predicted for a sample must exactly match the corresponding set of labels in y_test

Over here for solving the multilabel problem I have used the Binary Relevance method of Problem Transformation, i.e., I have divided the multi label problem into seperate single label problems.


### Multi-class and Multi-Label Classification Using Support Vector Machines- (b) (ii)

In [2]:
c = np.logspace(-4, 4, 9)    #weight of the SVM penalty
gam=np.linspace(0.01,2,9)    #width of the Gaussian Kernel
tuned_parameters = [{'gamma': gam,'C': c}]
svm_model=svm.SVC(kernel='rbf')   #Gaussian Radial Kernel-Radial Basis Function
gscv = GridSearchCV(svm_model, tuned_parameters, cv=10, n_jobs=-1)    #n_jobs for parallel processing 
gscv.fit(x_train,y_train['Family'])
print(gscv.best_params_)

{'C': 100.0, 'gamma': 2.0}


reference from https://scikit-learn.org/stable/auto_examples/model_selection/plot_grid_search_digits.html

In [3]:
clf=svm.SVC(C=10,gamma=2)
clf.fit(x_train,y_train['Family'])
y_pred1=clf.predict(x_test)
hl=hamming_loss(y_test['Family'], y_pred1)
ac=accuracy_score(y_test['Family'], y_pred1)
print("Hamming Loss with Family label=",hl)
print("Exact Match Score with Family label=",ac)

Hamming Loss with Family label= 0.004631773969430292
Exact Match Score with Family label= 0.9953682260305697


In [4]:
means = gscv.cv_results_['mean_test_score']
stds = gscv.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, gscv.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"% (mean, std * 2, params))

0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 0.01}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 0.25875}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 0.5075}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 0.75625}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 1.005}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 1.25375}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 1.5025}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 1.75125}
0.619 (+/-0.002) for {'C': 0.0001, 'gamma': 2.0}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 0.01}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 0.25875}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 0.5075}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 0.75625}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 1.005}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 1.25375}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 1.5025}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 1.75125}
0.619 (+/-0.002) for {'C': 0.001, 'gamma': 2.0}
0.619 (+/-0.002) for {'C': 0.01, 'gamma': 0.01}
0.823 (+/-0.025) for {'C': 0.

In [5]:
from sklearn.metrics import classification_report
print(classification_report(y_test['Family'], y_pred1))

              precision    recall  f1-score   support

           0       1.00      0.83      0.90        23
           1       1.00      1.00      1.00       177
           2       0.99      1.00      0.99       655
           3       1.00      1.00      1.00      1304

    accuracy                           1.00      2159
   macro avg       1.00      0.95      0.97      2159
weighted avg       1.00      1.00      1.00      2159



In [6]:
c = np.logspace(-4, 4, 9)
gam=np.linspace(0.01,2,9)
tuned_parameters = [{'gamma': gam,'C': c}]
svm_model=svm.SVC(kernel='rbf')   #Gaussian Radial Kernel-Radial Basis Function
gscv = GridSearchCV(svm_model, tuned_parameters, cv=10, n_jobs=-1)    #n_jobs for parallel processing 
gscv.fit(x_train,y_train['Genus'])
print(gscv.best_params_)

{'C': 10.0, 'gamma': 1.25375}


In [7]:
clf=svm.SVC(C=100,gamma=1.5025)
clf.fit(x_train,y_train['Genus'])
y_pred2=clf.predict(x_test)
hl=hamming_loss(y_test['Genus'], y_pred2)
ac=accuracy_score(y_test['Genus'], y_pred2)
print("Hamming Loss with Genus label=",hl)
print("Exact Match Score with Genus label=",ac)

Hamming Loss with Genus label= 0.007874015748031496
Exact Match Score with Genus label= 0.9921259842519685


In [8]:
means = gscv.cv_results_['mean_test_score']
stds = gscv.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, gscv.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"% (mean, std * 2, params))

0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 0.01}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 0.25875}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 0.5075}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 0.75625}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 1.005}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 1.25375}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 1.5025}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 1.75125}
0.580 (+/-0.005) for {'C': 0.0001, 'gamma': 2.0}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 0.01}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 0.25875}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 0.5075}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 0.75625}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 1.005}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 1.25375}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 1.5025}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 1.75125}
0.580 (+/-0.005) for {'C': 0.001, 'gamma': 2.0}
0.580 (+/-0.005) for {'C': 0.01, 'gamma': 0.01}
0.684 (+/-0.028) for {'C': 0.

In [9]:
print(classification_report(y_test['Genus'], y_pred2))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1230
           1       1.00      0.99      1.00       177
           2       0.96      0.97      0.97       101
           3       0.99      0.99      0.99       481
           4       0.99      0.99      0.99        74
           5       0.94      0.94      0.94        35
           6       1.00      0.83      0.90        23
           7       1.00      0.97      0.99        38

    accuracy                           0.99      2159
   macro avg       0.98      0.96      0.97      2159
weighted avg       0.99      0.99      0.99      2159



In [10]:
c = np.logspace(-4, 4, 9)
gam=np.linspace(0.01,2,9)
tuned_parameters = [{'gamma': gam,'C': c}]
svm_model=svm.SVC(kernel='rbf')   #Gaussian Radial Kernel-Radial Basis Function
gscv = GridSearchCV(svm_model, tuned_parameters, cv=10, n_jobs=-1)    #n_jobs for parallel processing 
gscv.fit(x_train,y_train['Species'])
print(gscv.best_params_)

{'C': 10.0, 'gamma': 1.25375}


In [11]:
clf=svm.SVC(C=10,gamma=2)
clf.fit(x_train,y_train['Species'])
y_pred3=clf.predict(x_test)
hl=hamming_loss(y_test['Species'], y_pred3)
ac=accuracy_score(y_test['Species'], y_pred3)
print("Hamming Loss with Species label=",hl)
print("Exact Match Score with Species label=",ac)

Hamming Loss with Species label= 0.006947660954145438
Exact Match Score with Species label= 0.9930523390458545


In [12]:
means = gscv.cv_results_['mean_test_score']
stds = gscv.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, gscv.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"% (mean, std * 2, params))

0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 0.01}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 0.25875}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 0.5075}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 0.75625}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 1.005}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 1.25375}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 1.5025}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 1.75125}
0.489 (+/-0.005) for {'C': 0.0001, 'gamma': 2.0}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 0.01}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 0.25875}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 0.5075}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 0.75625}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 1.005}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 1.25375}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 1.5025}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 1.75125}
0.489 (+/-0.005) for {'C': 0.001, 'gamma': 2.0}
0.489 (+/-0.005) for {'C': 0.01, 'gamma': 0.01}
0.642 (+/-0.009) for {'C': 0.

In [13]:
print(classification_report(y_test['Species'], y_pred3))

              precision    recall  f1-score   support

           0       0.99      1.00      0.99       216
           1       1.00      1.00      1.00      1014
           2       1.00      0.99      1.00       177
           3       0.97      0.98      0.98       101
           4       0.99      0.99      0.99       158
           5       0.99      1.00      0.99       323
           6       1.00      0.99      0.99        74
           7       0.97      0.94      0.96        35
           8       1.00      0.83      0.90        23
           9       0.97      0.97      0.97        38

    accuracy                           0.99      2159
   macro avg       0.99      0.97      0.98      2159
weighted avg       0.99      0.99      0.99      2159



In [14]:
y_true1=y_test['Family']
y_true2=y_test['Genus']
y_true3=y_test['Species']
pred_values=np.array(list(zip(y_pred1,y_pred2,y_pred3)))
true_values=np.array(list(zip(y_true1,y_true2,y_true3)))
hammin_sc=np.mean(pred_values==true_values)
hammin_loss=1-hammin_sc
print("Overall Hamming loss for Multilabel problem=",hammin_loss)
var=(pred_values==true_values)
exact_match=np.mean(np.product(var,axis=1))
print("Exact Match for the Multilabel Problem=",exact_match)

Overall Hamming loss for Multilabel problem= 0.006484483557202436
Exact Match for the Multilabel Problem= 0.9911996294580825


SVM:

Family Label:

Hamming Loss = 0.0046

Exact Match Score = 0.9954

Genus Label:

Hamming Loss = 0.0079

Exact Match Score = 0.9921

Species Label:

Hamming Loss = 0.0069

Exact Match Score = 0.9931

Overall Hamming loss for Multilabel problem= 0.0065

Exact Match for the Multilabel Problem= 0.9912

### Multi-class and Multi-Label Classification Using Support Vector Machines- (b) (iii)


In [15]:
from sklearn.preprocessing import StandardScaler
# Standardize to get zero mean and unit standard error
scaler=StandardScaler()
standard_train_x=scaler.fit_transform(x_train)  # fit and transform the data on the training set
standard_test_x=scaler.transform(x_test)        # apply the transformation on the test set


In [16]:
C = np.logspace(-4, 4, 9)
svm_linear_model=svm.LinearSVC(penalty='l1',dual=False)
tuned_paramater=dict(C=C)
gscv=GridSearchCV(svm_linear_model, tuned_paramater, cv=10, n_jobs=-1)
gscv.fit(standard_train_x,y_train['Family'])
print(gscv.best_params_)

{'C': 1.0}


In [17]:
clf=svm.LinearSVC(C=1)
clf.fit(standard_train_x,y_train['Family'])
ypred1=clf.predict(standard_test_x)
hl=hamming_loss(y_test['Family'], ypred1)
ac=accuracy_score(y_test['Family'], ypred1)
print("Hamming Loss with Family label=",hl)
print("Exact Match Score with Family label=",ac)

Hamming Loss with Family label= 0.0699397869383974
Exact Match Score with Family label= 0.9300602130616026


In [18]:
means = gscv.cv_results_['mean_test_score']
stds = gscv.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, gscv.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"% (mean, std * 2, params))

0.009 (+/-0.002) for {'C': 0.0001}
0.851 (+/-0.025) for {'C': 0.001}
0.927 (+/-0.025) for {'C': 0.01}
0.937 (+/-0.020) for {'C': 0.1}
0.939 (+/-0.019) for {'C': 1.0}
0.939 (+/-0.020) for {'C': 10.0}
0.938 (+/-0.020) for {'C': 100.0}
0.938 (+/-0.020) for {'C': 1000.0}
0.938 (+/-0.020) for {'C': 10000.0}


In [19]:
print(classification_report(y_test['Family'], ypred1))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00        23
           1       0.90      0.89      0.90       177
           2       0.91      0.89      0.90       655
           3       0.95      0.97      0.96      1304

    accuracy                           0.93      2159
   macro avg       0.69      0.69      0.69      2159
weighted avg       0.92      0.93      0.93      2159



In [20]:
C = np.logspace(-4, 4, 9)
svm_linear_model=svm.LinearSVC(penalty='l1',dual=False)
tuned_paramater=dict(C=C)
gscv=GridSearchCV(svm_linear_model, tuned_paramater, cv=10, n_jobs=-1)
gscv.fit(standard_train_x,y_train['Genus'])
print(gscv.best_params_)

{'C': 10.0}


In [21]:
clf=svm.LinearSVC(C=100)
clf.fit(standard_train_x,y_train['Genus'])
ypred2=clf.predict(standard_test_x)
hl=hamming_loss(y_test['Genus'], ypred2)
ac=accuracy_score(y_test['Genus'], ypred2)
print("Hamming Loss with Genus label=",hl)
print("Exact Match Score with Genus label=",ac)

Hamming Loss with Genus label= 0.05974988420565076
Exact Match Score with Genus label= 0.9402501157943493


In [22]:
means = gscv.cv_results_['mean_test_score']
stds = gscv.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, gscv.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"% (mean, std * 2, params))

0.580 (+/-0.005) for {'C': 0.0001}
0.818 (+/-0.017) for {'C': 0.001}
0.911 (+/-0.025) for {'C': 0.01}
0.938 (+/-0.018) for {'C': 0.1}
0.946 (+/-0.013) for {'C': 1.0}
0.947 (+/-0.013) for {'C': 10.0}
0.947 (+/-0.013) for {'C': 100.0}
0.947 (+/-0.012) for {'C': 1000.0}
0.947 (+/-0.012) for {'C': 10000.0}


In [23]:
print(classification_report(y_test['Genus'], ypred2))

              precision    recall  f1-score   support

           0       0.96      0.99      0.97      1230
           1       0.90      0.85      0.87       177
           2       0.75      0.64      0.69       101
           3       0.93      0.97      0.95       481
           4       0.95      0.82      0.88        74
           5       0.95      0.57      0.71        35
           6       0.90      0.83      0.86        23
           7       0.97      0.92      0.95        38

    accuracy                           0.94      2159
   macro avg       0.92      0.82      0.86      2159
weighted avg       0.94      0.94      0.94      2159



In [24]:
C = np.logspace(-4, 4, 9)
svm_linear_model=svm.LinearSVC(penalty='l1',dual=False)
tuned_paramater=dict(C=C)
gscv=GridSearchCV(svm_linear_model, tuned_paramater, cv=10, n_jobs=-1)
gscv.fit(standard_train_x,y_train['Species'])
print(gscv.best_params_)

{'C': 10.0}


In [25]:
clf=svm.LinearSVC(C=1)
clf.fit(standard_train_x,y_train['Species'])
ypred3=clf.predict(standard_test_x)
hl=hamming_loss(y_test['Species'], ypred3)
ac=accuracy_score(y_test['Species'], ypred3)
print("Hamming Loss with Species label=",hl)
print("Exact Match Score with Species label=",ac)

Hamming Loss with Species label= 0.04307549791570171
Exact Match Score with Species label= 0.9569245020842982


In [26]:
means = gscv.cv_results_['mean_test_score']
stds = gscv.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, gscv.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r"% (mean, std * 2, params))

0.091 (+/-0.001) for {'C': 0.0001}
0.780 (+/-0.022) for {'C': 0.001}
0.912 (+/-0.025) for {'C': 0.01}
0.948 (+/-0.019) for {'C': 0.1}
0.957 (+/-0.013) for {'C': 1.0}
0.958 (+/-0.013) for {'C': 10.0}
0.958 (+/-0.013) for {'C': 100.0}
0.958 (+/-0.013) for {'C': 1000.0}
0.958 (+/-0.013) for {'C': 10000.0}


In [27]:
print(classification_report(y_test['Species'], ypred3))

              precision    recall  f1-score   support

           0       0.94      0.94      0.94       216
           1       0.99      1.00      0.99      1014
           2       0.91      0.93      0.92       177
           3       0.90      0.77      0.83       101
           4       0.93      0.94      0.94       158
           5       0.92      0.98      0.95       323
           6       0.96      0.93      0.95        74
           7       1.00      0.57      0.73        35
           8       0.86      0.83      0.84        23
           9       0.95      0.95      0.95        38

    accuracy                           0.96      2159
   macro avg       0.94      0.88      0.90      2159
weighted avg       0.96      0.96      0.96      2159



In [28]:
pred_values=np.array(list(zip(ypred1,ypred2,ypred3)))
true_values=np.array(list(zip(y_true1,y_true2,y_true3)))
hammin_sc=np.mean(pred_values==true_values)
hammin_loss=1-hammin_sc
print("Overall Hamming loss for Multilabel problem=",hammin_loss)
var=(pred_values==true_values)
exact_match=np.mean(np.product(var,axis=1))
print("Exact Match for the Multilabel Problem=",exact_match)

Overall Hamming loss for Multilabel problem= 0.05758838968658331
Exact Match for the Multilabel Problem= 0.9018063918480778


L1-penalized SVM's:

Family Label:

Hamming Loss = 0.0699

Exact Match Score = 0.9301

Genus Label:

Hamming Loss = 0.0597

Exact Match Score = 0.9403

Species Label:

Hamming Loss = 0.0431

Exact Match Score = 0.9569

Overall Hamming loss for Multilabel problem= 0.0576

Exact Match for the Multilabel Problem= 0.9018

### Multi-class and Multi-Label Classification Using Support Vector Machines- (b) (iv)

In [29]:
! pip install imblearn

In [30]:
from collections import Counter
from imblearn.over_sampling import SMOTE
sm = SMOTE()
x_resampled, y_resampled = sm.fit_resample(standard_train_x, y_train['Family'])
print('Resampled dataset shape %s' % Counter(y_resampled))
print(y_resampled)
type(x_resampled)

Resampled dataset shape Counter({'2': 3116, '3': 3116, '1': 3116, '0': 3116})
['2' '3' '3' ... '2' '2' '2']


numpy.ndarray

In [31]:
C = np.logspace(-4, 4, 9)
svm_linear_model=svm.LinearSVC(penalty='l1',dual=False)
tuned_paramater=dict(C=C)
gscv=GridSearchCV(svm_linear_model, tuned_paramater, cv=10, n_jobs=-1)
gscv.fit(x_resampled,y_resampled)
print(gscv.best_params_)

{'C': 100.0}


In [32]:
clf=svm.LinearSVC(C=100)
clf.fit(standard_train_x,y_train['Family'])
y_pred11=clf.predict(standard_test_x)
hl=hamming_loss(y_test['Family'], y_pred11)
ac=accuracy_score(y_test['Family'], y_pred11)
print("Hamming Loss with Family label=",hl)
print("Exact Match Score with Family label=",ac)

Hamming Loss with Family label= 0.07688744789254284
Exact Match Score with Family label= 0.9231125521074571


In [33]:
x_resampled, y_resampled = sm.fit_resample(standard_train_x, y_train['Genus'])
print('Resampled dataset shape %s' % Counter(y_resampled))
print(y_resampled)
type(x_resampled)

Resampled dataset shape Counter({'3': 2920, '0': 2920, '4': 2920, '1': 2920, '2': 2920, '7': 2920, '5': 2920, '6': 2920})
['3' '0' '0' ... '7' '7' '7']


numpy.ndarray

In [34]:
C = np.logspace(-4, 4, 9)
svm_linear_model=svm.LinearSVC(penalty='l1',dual=False)
tuned_paramater=dict(C=C)
gscv=GridSearchCV(svm_linear_model, tuned_paramater, cv=10, n_jobs=-1)
gscv.fit(x_resampled,y_resampled)
print(gscv.best_params_)

{'C': 10.0}


In [35]:
clf=svm.LinearSVC(C=100)
clf.fit(standard_train_x,y_train['Genus'])
y_pred22=clf.predict(standard_test_x)
hl=hamming_loss(y_test['Genus'], y_pred22)
ac=accuracy_score(y_test['Genus'], y_pred22)
print("Hamming Loss with Genus label=",hl)
print("Exact Match Score with Genus label=",ac)

Hamming Loss with Genus label= 0.09726725335803613
Exact Match Score with Genus label= 0.9027327466419639


In [36]:
x_resampled, y_resampled = sm.fit_resample(standard_train_x, y_train['Species'])
print('Resampled dataset shape %s' % Counter(y_resampled))
print(y_resampled)
type(x_resampled)

Resampled dataset shape Counter({'5': 2464, '1': 2464, '6': 2464, '2': 2464, '0': 2464, '3': 2464, '4': 2464, '9': 2464, '7': 2464, '8': 2464})
['5' '1' '1' ... '9' '9' '9']


numpy.ndarray

In [37]:
C = np.logspace(-4, 4, 9)
svm_linear_model=svm.LinearSVC(penalty='l1',dual=False)
tuned_paramater=dict(C=C)
gscv=GridSearchCV(svm_linear_model, tuned_paramater, cv=10, n_jobs=-1)
gscv.fit(x_resampled,y_resampled)
print(gscv.best_params_)

{'C': 100.0}


In [38]:
clf=svm.LinearSVC(C=100)
clf.fit(standard_train_x,y_train['Species'])
y_pred33=clf.predict(standard_test_x)
hl=hamming_loss(y_test['Species'], y_pred33)
ac=accuracy_score(y_test['Species'], y_pred33)
print("Hamming Loss with Species label=",hl)
print("Exact Match Score with Species label=",ac)

Hamming Loss with Species label= 0.06252894858730894
Exact Match Score with Species label= 0.937471051412691


In [39]:
pred_values=np.array(list(zip(y_pred11,y_pred22,y_pred33)))
true_values=np.array(list(zip(y_true1,y_true2,y_true3)))
hammin_sc=np.mean(pred_values==true_values)
hammin_loss=1-hammin_sc
print("Overall Hamming loss for Multilabel problem=",hammin_loss)
var=(pred_values==true_values)
exact_match=np.mean(np.product(var,axis=1))
print("Exact Match for the Multilabel Problem=",exact_match)

Overall Hamming loss for Multilabel problem= 0.07889454994596268
Exact Match for the Multilabel Problem= 0.8360352014821677


SMOTE:

Family Label:

Hamming Loss = 0.0769

Exact Match Score = 0.9231

Genus Label:

Hamming Loss = 0.0973

Exact Match Score = 0.9027

Species Label:

Hamming Loss = 0.0625

Exact Match Score = 0.9375

Overall Hamming loss for Multilabel problem= 0.0789

Exact Match for the Multilabel Problem= 0.8360

From all the classifier results we can see that upsampling the data using smote gives us a lower accuracy score and higher hamming loss on the test data. We can also notice that SVM with the Gaussian Radial Kernel gives us a better result, than LinearSVM. 

In conclusion, the best method to use on this dataset is SVM with RBF Kernel followed by SVM with linear kernel and SVM with linear kernel using SMOTE for upsampling respectively. 

### 2. K-Means Clustering on a Multi-Class and Multi-Label Data Set Monte-Carlo Simulation (a)

Monte Carlo (MC) methods are a subset of computational algorithms that use the process of repeated random sampling to make numerical estimations of unknown parameters.

In [40]:
%%time
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score
silhouette_avg={}
list1=[]
for r in range(1,51):
    for k in range(2,51):
        k_means_model=KMeans(n_clusters=k,init='random',max_iter=100,n_jobs=-1)
        cluster_labels=k_means_model.fit_predict(data_x)
        silhouette_avg[k] = silhouette_score(data_x, cluster_labels)
#         print("For k =", k ,"The average silhouette_score is :", silhouette_score(data_x, cluster_labels))
    optimal_k = max(silhouette_avg,key=silhouette_avg.get)
    print("Optimal value of k for iterations {}={}".format(r,optimal_k))
    list1.append(optimal_k)

Optimal value of k for iterations 1=4
Optimal value of k for iterations 2=4
Optimal value of k for iterations 3=4
Optimal value of k for iterations 4=4
Optimal value of k for iterations 5=4
Optimal value of k for iterations 6=4
Optimal value of k for iterations 7=4
Optimal value of k for iterations 8=4
Optimal value of k for iterations 9=4
Optimal value of k for iterations 10=4
Optimal value of k for iterations 11=4
Optimal value of k for iterations 12=4
Optimal value of k for iterations 13=4
Optimal value of k for iterations 14=4
Optimal value of k for iterations 15=4
Optimal value of k for iterations 16=4
Optimal value of k for iterations 17=4
Optimal value of k for iterations 18=4
Optimal value of k for iterations 19=4
Optimal value of k for iterations 20=4
Optimal value of k for iterations 21=4
Optimal value of k for iterations 22=4
Optimal value of k for iterations 23=4
Optimal value of k for iterations 24=4
Optimal value of k for iterations 25=4
Optimal value of k for iterations 

I have taken the range from 2 to 50 and not 1 to 50 because number of labels is 1. Valid values are from 2 to n_samples - 1 (inclusive).

The best value of k(n_clusters) is the one which has the highest silhouette score which is at k=4.

The ideal value of si is 1. That means the higher the value of the silhouette score, the closer it is towards 1 and the data is very well clustered. The points in the cluster are very close to each other.

Monte Carlo reference from https://towardsdatascience.com/an-overview-of-monte-carlo-methods-675384eb1694

.get reference from https://stackoverflow.com/questions/26871866/print-highest-value-in-dict-with-key

KMeans reference from https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html

### 2. K-Means Clustering on a Multi-Class and Multi-Label Data Set Monte-Carlo Simulation (b)

In [237]:
df4=pd.DataFrame()
df_k=df.append(df4)
df_k['Family']=df_k['Family'].map({'0':'Bufonidae','1':'Dendrobatidae','2':'Hylidae','3':'Leptodactylidae'})
df_k['Genus']=df_k['Genus'].map({'0':'Adenomera','1':'Ameerega','2':'Dendropsophus','3':'Hypsiboas','4':'Leptodactylus',
                             '5':'Osteocephalus','6':'Rhinella','7':'Scinax'})
df_k['Species']=df_k['Species'].map({'0':'AdenomeraAndre','1':'AdenomeraHylaedactylus','2':'Ameeregatrivittata','3':'HylaMinuta',
                                '4':'HypsiboasCinerascens','5':'HypsiboasCordobae','6':'LeptodactylusFuscus',
                                 '7':'OsteocephalusOophagus','8':'Rhinellagranulosa','9':'ScinaxRuber'})

In [238]:
for k in list1:
    k_means=KMeans(n_clusters=k,init='random')
    model=k_means.fit(data_x)

    cls=np.unique(model.labels_)

    df_1 = pd.DataFrame()
    df_1['Index'] = df_k.index.values
    df_1['Cluster'] = model.labels_
    temp= df_1[df_1['Cluster']==0]['Index'].values    
    temp1=  df_1[df_1['Cluster']==1]['Index'].values
    temp2=  df_1[df_1['Cluster']==2]['Index'].values
    temp3=  df_1[df_1['Cluster']==3]['Index'].values
    
    f0=df_k.iloc[temp]['Family'].value_counts().idxmax()
    f1=df_k.iloc[temp1]['Family'].value_counts().idxmax()
    f2=df_k.iloc[temp2]['Family'].value_counts().idxmax()
    f3=df_k.iloc[temp3]['Family'].value_counts().idxmax()
    g0=df_k.iloc[temp]['Genus'].value_counts().idxmax()
    g1=df_k.iloc[temp1]['Genus'].value_counts().idxmax()
    g2=df_k.iloc[temp2]['Genus'].value_counts().idxmax()
    g3=df_k.iloc[temp3]['Genus'].value_counts().idxmax()
    s0=df_k.iloc[temp]['Species'].value_counts().idxmax()
    s1=df_k.iloc[temp1]['Species'].value_counts().idxmax()
    s2=df_k.iloc[temp2]['Species'].value_counts().idxmax()
    s3=df_k.iloc[temp3]['Species'].value_counts().idxmax()   
    
    print("Majority Class in Family Label: \n Cluster 0:{}, Cluster 1:{}, Cluster 2:{}, Cluster 3:{}".format(f0,f1,f2,f3))
    print("Majority Class in Genus Label: \n Cluster 0:{}, Cluster 1:{}, Cluster 2:{}, Cluster 3:{}".format(g0,g1,g2,g3))
    print("Majority Class in Species Label: \n Cluster 0:{}, Cluster 1:{}, Cluster 2:{}, Cluster 3:{}".format(s0,s1,s2,s3))

Majority Class in Family Label: 
 Cluster 0:Leptodactylidae, Cluster 1:Hylidae, Cluster 2:Hylidae, Cluster 3:Leptodactylidae
Majority Class in Genus Label: 
 Cluster 0:Adenomera, Cluster 1:Hypsiboas, Cluster 2:Hypsiboas, Cluster 3:Adenomera
Majority Class in Species Label: 
 Cluster 0:AdenomeraAndre, Cluster 1:HypsiboasCordobae, Cluster 2:HypsiboasCinerascens, Cluster 3:AdenomeraHylaedactylus
Majority Class in Family Label: 
 Cluster 0:Dendrobatidae, Cluster 1:Leptodactylidae, Cluster 2:Hylidae, Cluster 3:Hylidae
Majority Class in Genus Label: 
 Cluster 0:Ameerega, Cluster 1:Adenomera, Cluster 2:Hypsiboas, Cluster 3:Hypsiboas
Majority Class in Species Label: 
 Cluster 0:Ameeregatrivittata, Cluster 1:AdenomeraHylaedactylus, Cluster 2:HypsiboasCinerascens, Cluster 3:HypsiboasCordobae
Majority Class in Family Label: 
 Cluster 0:Leptodactylidae, Cluster 1:Dendrobatidae, Cluster 2:Hylidae, Cluster 3:Hylidae
Majority Class in Genus Label: 
 Cluster 0:Adenomera, Cluster 1:Ameerega, Cluster 2:

Majority Class in Family Label: 
 Cluster 0:Hylidae, Cluster 1:Leptodactylidae, Cluster 2:Hylidae, Cluster 3:Dendrobatidae
Majority Class in Genus Label: 
 Cluster 0:Hypsiboas, Cluster 1:Adenomera, Cluster 2:Hypsiboas, Cluster 3:Ameerega
Majority Class in Species Label: 
 Cluster 0:HypsiboasCordobae, Cluster 1:AdenomeraHylaedactylus, Cluster 2:HypsiboasCinerascens, Cluster 3:Ameeregatrivittata
Majority Class in Family Label: 
 Cluster 0:Hylidae, Cluster 1:Hylidae, Cluster 2:Dendrobatidae, Cluster 3:Leptodactylidae
Majority Class in Genus Label: 
 Cluster 0:Hypsiboas, Cluster 1:Hypsiboas, Cluster 2:Ameerega, Cluster 3:Adenomera
Majority Class in Species Label: 
 Cluster 0:HypsiboasCordobae, Cluster 1:HypsiboasCinerascens, Cluster 2:Ameeregatrivittata, Cluster 3:AdenomeraHylaedactylus
Majority Class in Family Label: 
 Cluster 0:Hylidae, Cluster 1:Leptodactylidae, Cluster 2:Leptodactylidae, Cluster 3:Hylidae
Majority Class in Genus Label: 
 Cluster 0:Hypsiboas, Cluster 1:Adenomera, Cluste

Majority Class in Family Label: 
 Cluster 0:Hylidae, Cluster 1:Leptodactylidae, Cluster 2:Hylidae, Cluster 3:Leptodactylidae
Majority Class in Genus Label: 
 Cluster 0:Hypsiboas, Cluster 1:Adenomera, Cluster 2:Hypsiboas, Cluster 3:Adenomera
Majority Class in Species Label: 
 Cluster 0:HypsiboasCordobae, Cluster 1:AdenomeraHylaedactylus, Cluster 2:HypsiboasCinerascens, Cluster 3:AdenomeraAndre
Majority Class in Family Label: 
 Cluster 0:Hylidae, Cluster 1:Hylidae, Cluster 2:Leptodactylidae, Cluster 3:Dendrobatidae
Majority Class in Genus Label: 
 Cluster 0:Hypsiboas, Cluster 1:Hypsiboas, Cluster 2:Adenomera, Cluster 3:Ameerega
Majority Class in Species Label: 
 Cluster 0:HypsiboasCordobae, Cluster 1:HypsiboasCinerascens, Cluster 2:AdenomeraHylaedactylus, Cluster 3:Ameeregatrivittata
Majority Class in Family Label: 
 Cluster 0:Leptodactylidae, Cluster 1:Leptodactylidae, Cluster 2:Hylidae, Cluster 3:Hylidae
Majority Class in Genus Label: 
 Cluster 0:Adenomera, Cluster 1:Adenomera, Cluster

### 2. K-Means Clustering on a Multi-Class and Multi-Label Data Set Monte-Carlo Simulation (c)

The Hamming distance between two strings of equal length is the number of positions at which the corresponding symbols are different. This means that Hamming distance is equal to Hamming Loss. 

In [250]:
i =1
list_hammin=[]
list_loss=[]
for k in list1:
    k_means=KMeans(n_clusters=k,init='random')
    model=k_means.fit(data_x)
    df_1 = pd.DataFrame()
    df_1['Index'] = df_k.index.values
    df_1['Cluster'] = model.labels_
    temp= df_1[df_1['Cluster']==0]['Index'].values    
    temp1=  df_1[df_1['Cluster']==1]['Index'].values
    temp2=  df_1[df_1['Cluster']==2]['Index'].values
    temp3=  df_1[df_1['Cluster']==3]['Index'].values
    
    f0=df_k.iloc[temp]['Family'].value_counts().idxmax()
    f1=df_k.iloc[temp1]['Family'].value_counts().idxmax()
    f2=df_k.iloc[temp2]['Family'].value_counts().idxmax()
    f3=df_k.iloc[temp3]['Family'].value_counts().idxmax()
    g0=df_k.iloc[temp]['Genus'].value_counts().idxmax()
    g1=df_k.iloc[temp1]['Genus'].value_counts().idxmax()
    g2=df_k.iloc[temp2]['Genus'].value_counts().idxmax()
    g3=df_k.iloc[temp3]['Genus'].value_counts().idxmax()
    s0=df_k.iloc[temp]['Species'].value_counts().idxmax()
    s1=df_k.iloc[temp1]['Species'].value_counts().idxmax()
    s2=df_k.iloc[temp2]['Species'].value_counts().idxmax()
    s3=df_k.iloc[temp3]['Species'].value_counts().idxmax()   
    
    print("Majority Triplets for iteration {}: \n Cluster 0:{},{},{} \n Cluster 1: {},{},{} \n Cluster 2: {},{},{} \n Cluster 3: {},{},{} ".format(i,f0,g0,s0,f1,g1,s1,f2,g2,s2,f3,g3,s3))
    df_2=pd.DataFrame()
    df_3=df_k.append(df_2)
    df_3.loc[temp,'Family']=df_k.iloc[temp]['Family'].value_counts().idxmax()
    df_3.loc[temp1,'Family']=df_k.iloc[temp1]['Family'].value_counts().idxmax()
    df_3.loc[temp2,'Family']=df_k.iloc[temp2]['Family'].value_counts().idxmax()
    df_3.loc[temp3,'Family']=df_k.iloc[temp3]['Family'].value_counts().idxmax()
    df_3.loc[temp,'Genus']=df_k.iloc[temp]['Genus'].value_counts().idxmax()
    df_3.loc[temp1,'Genus']=df_k.iloc[temp1]['Genus'].value_counts().idxmax()
    df_3.loc[temp2,'Genus']=df_k.iloc[temp2]['Genus'].value_counts().idxmax()
    df_3.loc[temp3,'Genus']=df_k.iloc[temp3]['Genus'].value_counts().idxmax()
    df_3.loc[temp,'Species']=df_k.iloc[temp]['Species'].value_counts().idxmax()
    df_3.loc[temp1,'Species']=df_k.iloc[temp1]['Species'].value_counts().idxmax()
    df_3.loc[temp2,'Species']=df_k.iloc[temp2]['Species'].value_counts().idxmax()
    df_3.loc[temp3,'Species']=df_k.iloc[temp3]['Species'].value_counts().idxmax()
    label_true1=df_k['Family']
    label_true2=df_k['Genus']
    label_true3=df_k['Species']
    label_pred1=df_3['Family']
    label_pred2=df_3['Genus']
    label_pred3=df_3['Species']
    pred_values=np.array(list(zip(label_pred1,label_pred2,label_pred3)))
    true_values=np.array(list(zip(label_true1,label_true2,label_true3)))
    hammin_sc=np.mean(pred_values==true_values) 
    hammin_loss=1-hammin_sc
    print("Hamming score for iteration {}={}".format(i,hammin_sc))
    print("Hamming loss for iteration {}={}".format(i,hammin_loss))
    list_loss.append(hammin_loss)
    list_hammin.append(hammin_sc)
    i+=1

Majority Triplets for iteration 1: 
 Cluster 0:Dendrobatidae,Ameerega,Ameeregatrivittata 
 Cluster 1: Hylidae,Hypsiboas,HypsiboasCinerascens 
 Cluster 2: Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 3: Hylidae,Hypsiboas,HypsiboasCordobae 
Hamming score for iteration 1=0.7775770210794534
Hamming loss for iteration 1=0.22242297892054663
Majority Triplets for iteration 2: 
 Cluster 0:Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 1: Dendrobatidae,Ameerega,Ameeregatrivittata 
 Cluster 2: Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 3: Hylidae,Hypsiboas,HypsiboasCinerascens 
Hamming score for iteration 2=0.7775770210794534
Hamming loss for iteration 2=0.22242297892054663
Majority Triplets for iteration 3: 
 Cluster 0:Leptodactylidae,Adenomera,AdenomeraAndre 
 Cluster 1: Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 2: Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 3: Hylidae,Hypsiboas,HypsiboasCordobae 
Hamming score for iteration 3=0.7524669909659486


Majority Triplets for iteration 25: 
 Cluster 0:Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 1: Hylidae,Hypsiboas,HypsiboasCinerascens 
 Cluster 2: Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 3: Dendrobatidae,Ameerega,Ameeregatrivittata 
Hamming score for iteration 25=0.7775770210794534
Hamming loss for iteration 25=0.22242297892054663
Majority Triplets for iteration 26: 
 Cluster 0:Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 1: Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 2: Hylidae,Hypsiboas,HypsiboasCinerascens 
 Cluster 3: Dendrobatidae,Ameerega,Ameeregatrivittata 
Hamming score for iteration 26=0.7775770210794534
Hamming loss for iteration 26=0.22242297892054663
Majority Triplets for iteration 27: 
 Cluster 0:Dendrobatidae,Ameerega,Ameeregatrivittata 
 Cluster 1: Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 2: Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 3: Hylidae,Hypsiboas,HypsiboasCinerascens 
Hamming score for iteration 27=0.77757

Majority Triplets for iteration 49: 
 Cluster 0:Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 1: Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
 Cluster 2: Dendrobatidae,Ameerega,Ameeregatrivittata 
 Cluster 3: Hylidae,Hypsiboas,HypsiboasCinerascens 
Hamming score for iteration 49=0.7780866342367385
Hamming loss for iteration 49=0.22191336576326148
Majority Triplets for iteration 50: 
 Cluster 0:Hylidae,Hypsiboas,HypsiboasCordobae 
 Cluster 1: Dendrobatidae,Ameerega,Ameeregatrivittata 
 Cluster 2: Hylidae,Hypsiboas,HypsiboasCinerascens 
 Cluster 3: Leptodactylidae,Adenomera,AdenomeraHylaedactylus 
Hamming score for iteration 50=0.7775770210794534
Hamming loss for iteration 50=0.22242297892054663


Replacing values based on index reference from https://stackoverflow.com/questions/37725195/pandas-replace-values-based-on-index

Hamming Distance reference from https://en.wikipedia.org/wiki/Hamming_distance

In [251]:
print("Average Hamming Score of 50 iterations",np.mean(list_hammin))
print("Average Hamming Distance of 50 iterations",np.mean(list_loss))
print("Standard Deviation of Hamming Distance of 50 iterations",np.std(list_loss))

Average Hamming Score of 50 iterations 0.7725855918461895
Average Hamming Distance of 50 iterations 0.22741440815381053
Standard Deviation of Hamming Distance of 50 iterations 0.008992983804757201
